### DISCUSSION TITLE
Understanding use of assets across code locations with Definitions

### DISCUSSION THREAD

**Discussion by mcenno** [dagster-io/dagster#11520] [Q&A] [answered]
I have learned about Dagster only yesterday and am tinkering with it to see if it fits my needs. One use case is to read a config file for an ETL procedure involving many different data files, and to provide the config to each subsequent step. I discovered I can do this with `Definitions` and `SourceAsset`, but am encountering an error I do not understand. Here is a minimum working example, consisting of four files:

`workspace.yaml` defines which files to use:

```
load_from:
  - python_file: file1.py
  - python_file: file2.py
  - python_file: file3.py
```

`file1.py` contains the configuration data, which is to be fed into subsequent processing steps:
```
from dagster import asset, Definitions

@asset
def config_data():
    return 1

defs = Definitions(assets=[config_data])
```

`file2.py` imports the configuration asset and provides two more assets:
```
from dagster import (asset, Definitions, SourceAsset, AssetKey)

config_data = SourceAsset(key=AssetKey("config_data"))

@asset
def file2_asset1(config_data):
    return config_data + 1

@asset
def file2_asset2(config_data):
    return config_data + 2

defs = Definitions(assets=[file2_asset1, file2_asset2])
```

`file3.py` uses assets from `file1.py` and `file2.py`:
```
from dagster import (asset, SourceAsset, AssetKey)

config_data = SourceAsset(key=AssetKey("config_data"))
file2_asset1 = SourceAsset(key=AssetKey("file2_asset1"))
file2_asset2 = SourceAsset(key=AssetKey("file2_asset2"))

@asset
def file3_asset1(config_data, file2_asset1, file2_asset2):
    return config_data + file2_asset1 + file2_asset2 + 1
```

The code as shown above does not work, but results in an error:
`DagsterInvalidDefinitionError: Input asset '["config_data"]' for asset '["file2_asset1"]' is not produced by any of the provided asset ops and is not one of the provided sources`.  However, if I add `config_data` to the `Definitions` at the end of `file2.py`, it works as expected.

Why does this happen and how can I solve this? Is there another / better way to provide assets across multiple files?

Any insights are much appreciated.

**Comment by schrockn** [comment, marked as answer]
Thanks for the question @mcenno. The issue here is that you need pass the SourceAsset to the Definitions object:

```python
from dagster import (asset, Definitions, SourceAsset, AssetKey)

config_data = SourceAsset(key=AssetKey("config_data"))

@asset
def file2_asset1(config_data):
    return config_data + 1

@asset
def file2_asset2(config_data):
    return config_data + 2

defs = Definitions(assets=[config_data, file2_asset1, file2_asset2]) # must pass config_data here
```

**Reply by mcenno** [reply]
Does that mean that I cannot use an asset multiple times without including it in each and every `Definitions` every time I use it?

**Reply by schrockn** [reply]
Yes. Each of those code locations is a self-contained namespace in its own process, so each one needs to know about all the assets it refers to.

What is your goals in using multiple Definitions objects in this case? Organizing your definitions in the UI? Do the assets use different dependencies?

**Reply by mcenno** [reply]
Thanks for your reply. I want to define many assets in multiple files to keep the files short and tidy. There are many dependencies between the assets, for example, the configuration file will be needed in almost all of them, and some assets are used frequently, ie, many other assets need them to compute their results.

I was expecting I could simply provide the relevant "inpput" assets to a file, compute something, and use `Definitions` to provide the result as "output" to other assets - just like I was trying to demonstrate in my example. But that is apparently not how it works.

**Reply by sryza** [reply]
@mcenno I think the most typical way to use Dagster for the situation you're describing would be to have a single code location and single `Definitions` object that includes all of those files.

The main reason to use separate code locations if you want your code to live in different Python environments (e.g. with different dependencies and different containers). If your goal is just to organize code in different files, they should probably live in the same code location.

You can use `group_name` to organize how your assets show up in the UI.

**Comment by andreapiso** [comment]
Does that mean we cannot set asset lineage across different code locations? Let's say I have two independent teams - a data engineering team and a data science team. They each use a code location as they work separately and we want to isolate their work. However, the data science team needs the assets built by the data engineering team (living in a different code location) to run their ML workloads.

This feels like it would be a reasonable usage pattern, and airflow would be able to support it with cross-DAG dependencies. Would I be able to achieve this with Dagster?

**Reply by andreapiso** [reply]
Looking at the example - looks like this actually can be achieved! Simply need to first add the definition - even if the asset is in a different code location. Correct?

**Reply by sryza** [reply]
Asset dependencies can span code locations.

If you're using IO managers, you need to have a `SourceAsset` in the downstream location that corresponds to the asset in the upstream location. The `SourceAsset` contains the metadata that tells Dagster how to load the assets that are downstream of it.
